File: C++/ps_baekjoon/1525/main.cpp
#include <bits/stdc++.h>
using namespace std;

constexpr int SOLVED = 123456789, SIZE = 3, dr[]{0, 0, +1, -1}, dc[]{+1, -1, 0, 0};
int puzzle[SIZE][SIZE];

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    string str;
    str.reserve(9);
    for (int i = 0; i < 9; ++i)
    {
        int num;
        cin >> num;
        if (num == 0)
            num = 9;
        str.push_back('0' + num);
    }
    int init = stoi(str);
    queue<int> q;
    unordered_map<int, int> vst;
    q.push(init);
    vst[init] = 0;
    while (!q.empty())
    {
        int now_int = q.front();
        string now_str = to_string(now_int);
        q.pop();
        if (now_int == SOLVED)
            break;
        int pos = now_str.find('9');
        int r = pos / 3;
        int c = pos % 3;
        for (int direction = 0; direction < 4; ++direction)
        {
            int nr = r + dr[direction];
            int nc = c + dc[direction];
            if (!(0 <= nr && nr < 3 && 0 <= nc && nc < 3))
                continue;
            string temp = now_str;
            swap(temp[pos], temp[nr * 3 + nc]);
            int next = stoi(temp);
            if (!vst[next])
            {
                vst[next] = vst[now_int] + 1;
                q.push(next);
            }
        }
    }
    cout << (vst.find(SOLVED) == vst.end() ? -1 : vst[SOLVED]);
    return 0;
}

File: Python/ps_baekjoon/1525/main.py
import sys
from queue import Queue
from collections import defaultdict

SOLVED = "123456789"
SIZE = 3

puzzle = [list(sys.stdin.readline().replace("0", "9").strip().split(" "))[:SIZE] for _ in range(SIZE)]
init = "".join(e for inner in puzzle for e in inner)
q = Queue()
vst = defaultdict(int)
vst[init] = 0
q.put(init)
while not q.empty():
    now = q.get()
    if now == SOLVED:
        break
    idx = now.find("9")
    r, c = [idx // 3, idx % 3]
    for dr, dc in zip([0, 0, +1, -1], [+1, -1, 0, 0]):
        nr, nc = [r + dr, c + dc]
        if not (0 <= nr < 3 and 0 <= nc < 3):
            continue
        temp = now.replace(now[idx], "-").replace(now[nr * 3 + nc], now[idx]).replace("-", now[nr * 3 + nc])
        if not vst[temp]:
            vst[temp] = vst[now] + 1
            q.put(temp)
print(vst[SOLVED] if SOLVED in vst else -1)